proposed/2023/NuGetAudit-supress-advisory.md

```diff
@@ -0,0 +1,231 @@
+
+Additionally, this will only work for direct packages, not transitive packages.
+If we use this approach for direct packages, there's no obvious alternative for transitive packages.
+Using a new MSBuild item for transitive packages, but metadata on a PackageReference item for direct packages is confusing because developers need more context in order to choose the correct syntax.
+It also means duplication when a package is direct in one project, but transitive in others.
+Having the suppress metadata flowed down to transitive packages, as Include/Exclude/Private Assets does is another option, because it's behavior that already exists.
+However, this makes it difficult to suppress centrally, like in a *Directory.Packages.props* file, as a transitive package will have different "entry points" in different projects in a solution.
+
+### Performance concerns
+
+Restore is NuGet's most performance critical path.
+Generally, .NET application performance is hurt most by memory allocations, and using MSBuild items is probably the worst choice for memory allocations.
+
+If the setting was in nuget.config, it would be loaded only once, and then a single instance would be passed around in memory when restoring multiple projects in parallel.
+In MSBuild, items are an object, probably causing multiple allocations, whereas a property is a single string.
+Unless MSBuild implements copy-on-write when modifying MSBuild items, it's likely that MSBuild will allocate a duplicate for every project that imports a shared props file.
+Even if we used a property, it's likely that MSBuild would allocate a duplicate string in each project importing a shared props file.
+Even if MSBuild did smart string pooling, when Nuget reads a dgspec file, since the dgspec is per-project, NuGet would need to do string pooling across parallel project reads to deduplicate those strings.
+So, it's very likely that using MSBuild at all will cause many allocations, both in MSBuild (and therefore when NuGet isn't even run), as well as within NuGet itself.
+
+Additionally, by using MSBuild items and properties, it forces MSBuild to do these allocations on every project load, even when restore (or `dotnet list package`) is not run.
+Putting it in *nuget.config* would not only make it much easier to avoid duplicate strings, it would also only be read when NuGet is actually used vs every MSBuild project evaluation.
+
+However, basic tests on MSBuild evaluation performance doesn't show measurable regressions when adding 100 additional items into a project, and I don't believe it's likely that customers will suppress 100 different advisories.
+Therefore, even though using MSBuild items is theoretically the highest allocation design choice, given how much other work MSBuild does, it doesn't appear to result in a measurable performance impact.
+
+## Prior Art
+
+<!-- What prior art, both good and bad are related to this proposal? -->
```

Comment:
When thinking about prior art we can draw inspiration from, consider GitHub dependabot suppressions, CG suppressions, NPM. It will be interesting to see the granularity that each solution provides (per package, per project, per advisory, etc.). Assuming that NuGet users already have experience using those other vuln suppression mechanism, they likely expect a similar customer experience from NuGet.